transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java

```diff
@@ -0,0 +1,250 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.AbstractChannel;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelMetadata;
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.EventLoop;
+import io.netty.channel.RecvByteBufAllocator;
+import io.netty.channel.unix.UnixChannel;
+import io.netty.channel.unix.UnixChannelUtil;
+import io.netty.util.ReferenceCountUtil;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.channels.UnresolvedAddressException;
+
+import static io.netty.util.internal.ObjectUtil.*;
+
+public abstract class AbstractIOUringChannel extends AbstractChannel implements UnixChannel {
+    private volatile SocketAddress local;
+    private static final ChannelMetadata METADATA = new ChannelMetadata(false);
+    final LinuxSocket socket;
+    protected volatile boolean active;
+    boolean uringInReadyPending;
+
+    AbstractIOUringChannel(final Channel parent, LinuxSocket fd) {
+        super(parent);
+        this.socket = checkNotNull(fd, "fd");
+        this.active = true;
+    }
+
+    public boolean isOpen() {
+        return socket.isOpen();
+    }
+
+    @Override
+    public boolean isActive() {
+        return active;
+    }
+
+    @Override
+    public ChannelMetadata metadata() {
+        return METADATA;
+    }
+
+    @Override
+    protected abstract AbstractUringUnsafe newUnsafe();
+
+    @Override
+    protected boolean isCompatible(final EventLoop loop) {
+        return loop instanceof IOUringEventLoop;
+    }
+
+    public void doReadBytes(ByteBuf byteBuf) {
+        IOUringEventLoop ioUringEventLoop = (IOUringEventLoop) eventLoop();
+        IOUringSubmissionQueue submissionQueue = ioUringEventLoop.getRingBuffer().getIoUringSubmissionQueue();
+
+        unsafe().recvBufAllocHandle().attemptedBytesRead(byteBuf.writableBytes());
+
+        if (byteBuf.hasMemoryAddress()) {
+            long eventId = ioUringEventLoop.incrementEventIdCounter();
+            final Event event = new Event();
+            event.setId(eventId);
+            event.setOp(EventType.READ);
+            event.setReadBuffer(byteBuf);
+            event.setAbstractIOUringChannel(this);
+            submissionQueue.add(eventId, EventType.READ, socket.getFd(), byteBuf.memoryAddress(),
+                                byteBuf.writerIndex(), byteBuf.capacity());
+        }
```

Comment:
the idea is throw it in `submissionQueue.add` for example if the submissionQueue is full

Reply:
sorry I don't get what you are saying... To me it seems like if we don't have a memoryAddress we will never be able to make progress so we need to throw. If what you are saying is that this will never happen we should add at least an assert and add a comment.

Reply:
there are two related problems
`io_uring_enter` can return https://www.mankier.com/2/io_uring_enter#EBUSY but we can avoid if we keep track of events and use kernel mask which means that https://github.com/1Jo1/netty/blob/io-uring/transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java#L81
it is possible that existing events which are not executed are overwritten

Reply:
oh yeah good point, we should add assert & comment for the memoryAddress